aiozk/recipes/base_lock.py

```diff
@@ -2,17 +2,30 @@
 import logging
 
 from aiozk import exc, states, Deadline
-
 from .sequential import SequentialRecipe
 
 log = logging.getLogger(__name__)
 
 
 class BaseLock(SequentialRecipe):
-    async def wait_in_line(self, znode_label, timeout=None, blocked_by=None):
+    def __init__(self, base_path):
+        super().__init__(base_path)
+        self._session_lost_future = None
+        self._alert_task = None
+
+    async def alert_session_lost(self):
+        await self._session_lost_future
+        log.warning(
+            'Session expired at some point, lock %s no longer acquired.', self)
+
+    async def _wait_in_line(self,
+                            znode_label,
+                            timeout=None,
+                            wait_with_party=False):
```

Comment:
`blocked_by` was more explicit in two aspects:
1. users understand that we need to wait for another lock
2. users understand what exactly we're waiting for

In general, I don't think that `wait_in_line` really the best function that we should use. Probably we should have some function that allows to wait for lock without acquiring it. So read lock would be constructed like:

```python
read_lock = Lock('read', blocked_by=[Lock('write', only_wait=True)]
# or with | or &
read_lock = Lock('read') | Lock('write', only_wait=True)
```

Reply:
OK. `blocked_by` will be returned.

However I do not understand the merit of the design you suggested.
IMO instantiating Lock objects and combining them may provide higher abstraction but I don't think it is needed now.
Maybe we can conduct refactoring later.
Also operator overloading does not catch my mind. It makes code be more complicated. And hard to get along with `grep`

Reply:
In that case, I just want to make it more obvious that we're interacting with another `Lock`.

Operator overloading is probably too much, but ideally, it would be great if we can hide some low-level details with better abstractions. For example, lock code itself has only cryptic `wait_in_line` and `wait_in_line` has quite many repetitions of `while True: .. except SessionLost` pattern so it is hard to follow. I'm not ready to give a better design yet.

We definitely will merge this PR without major changes. But if you have any ideas about making code better - probably it is a good time to share them.